`tree-wide`: add base `kafka::log_reader_config`

WillemKauf opened the pull request:
Up till now, the `storage::log_reader_config` has been used as the way to specify the parameters by which to read from a Kafka log in `redpanda`. Unfortunately, this is a leaky abstraction, as it contains some local `storage` specific flags that don't have any meaning outside of a `log_reader` created over local `segment`s.

Despite this, `storage::log_reader_config` has been re-used in `cloud_storage` code, as well as newly introduced `cloud_topics` code. 

This PR introduces:
1. A base `kafka::log_reader_config` struct which contains only fundamental types agnostic to the underlying data source, intended to be embedded within the other more subsystem specific objects.
2. A `storage::local_log_reader_config` struct which represents the configuration for a read from the local log.
3. A `cloud_storage::cloud_log_reader_config` struct which represents the configuration for a read from tiered storage.
4. A `cloud_topics::level_zero_log_reader_config` struct which represents the configuration for a read from L0 data in `cloud_topics`.

## Backports Required

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

WillemKauf committed:
`kafka`: add `log_reader_config`

As a base level abstraction to represent the desire to read the `log`
per certain inputs such as offset range, maximum bytes, etc.

This will be used in future commits, the idea being log reader configs
for specific data sources (local storage, tiered storage, cloud topics, etc)
will have specific objects that wrap an underlying `kafka::log_reader_config`,
and potentially contain additional flags/state.

WillemKauf committed:
`model`: move `using` definitions to `fundamental.h`

WillemKauf committed:
`cloud_storage`: add `cloud_log_reader_config`

nvartolomei commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,139 @@
+    size_t min_bytes;
+    size_t max_bytes;
+
+    std::optional<model::record_batch_type> type_filter;
```
This does not make sense in the kafka namespace. Is it possible to (re)move it? From "kafka" point of view there are only one type of batches (at the storage level they are raft data batches). I reckon this is completely unnecessary for cloud topics (i don't know the latest updates on the spec there). We're leaking abstractions here.

WillemKauf replied:
I shall take a look at moving this to a lower level `reader_config` :+1:

nvartolomei commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,139 @@
+    model::opt_abort_source_t abort_source;
+
+    // used by log reader
+    size_t bytes_consumed{0};
```
Why a config would know about bytes_consumed? If an implementation needs to track that then they should do it there. Why should the caller provide the (memory) space for that?

WillemKauf replied:
Once again I'll look at moving this to a lower level reader config, it (as the comment describes) is used entirely for book keeping within the `log_reader` implementation.

nvartolomei commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,139 @@
+    // skipping consumer sets this bit if a read would cause max_bytes to be
+    // violated.  its used to signal to the log reader that even though consumed
+    // bytes hasn't reached max bytes that reading should still stop.
+    bool over_budget{false};
```
Ditto. Why a particular consumer implementation detail leaks into a (kafka level) config?

WillemKauf replied:
Moved to lower level reader configs

nvartolomei commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,139 @@
+    // The assumption here is that the gap was created by compaction, and that
+    // the corresponding ghost batches exactly fill the space, while preserving
+    // terms boundaries.
+    bool fill_gaps{false};
```
Another abstraction leaks. No kafka level abstraction has a use case for filling the gaps and even if it did it would be a no-op since kafka level abstractions are only interested in data batches anyway and gaps are filled with ghost batches.

WillemKauf replied:
Moved to lower level reader configs

nvartolomei commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,139 @@
+    // NOTE: the translation refers only to the returned batches, not to the
+    // input min/max offset bounds. Callers are expected to account for inputs
+    // separately.
+    translate_offsets translate_offsets{false};
```
Ditto. From kafka point of view there is always only one type of offsets (kafka aka translated). This doesn't belong here.

nvartolomei commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,139 @@
+
+// Empty, invalid reader config which is sometimes useful as a placeholder
+// since log_reader_config doesn't have a default constructor.
+static const log_reader_config empty_reader_config{{}, {}};
```
Didn't yet read through the use-case of this but I believe this should be marked inline

WillemKauf replied:
Marked as `inline`

nvartolomei commented on src/v/cloud_topics/level_zero/reader/reader.h:
```diff
@@ -20,6 +20,20 @@ class partition;
 class data_plane_api;
 
+struct level_zero_log_reader_config {
+    static level_zero_log_reader_config
```
wonder why not explicit constructor instead?

WillemKauf replied:
`explicit` constructor would also be a good option here, no qualms about changing construction away from this annoying-to-type `static` factory function.

WillemKauf replied:
Changed to `explicit` constructor and removed `static` factory functions.

nvartolomei commented on src/v/cloud_topics/level_zero/reader/reader.h:
```diff
@@ -20,6 +20,20 @@ class partition;
+        return level_zero_log_reader_config{.cfg = std::move(cfg)};
+    }
+
+    kafka::log_reader_config cfg;
```
nit: maybe call it `kcfg` so that when it is reference as cfg.cfg or config.cfg it is less "abstract". `cfg.kcfg` reads slightly better to me

nvartolomei reviewed (commented):
I like the idea but I think the execution makes it even more confusing/unreadable than it was before. We're moving things around, add extra classes but we don't break any logical dependency chains.

**Anyway, what values does this change bring?** Even if incremental. We're still leaking abstractions all over, we're still mutating and re-using model::offsets to refer to both kafka offsets and raft offsets depending on when you inspect the config.

WillemKauf commented:
> I like the idea but I think the execution makes it even more confusing/unreadable than it was before. We're moving things around, add extra classes but we don't break any logical dependency chains.

I agree that it turned out a bit less than beautiful (ugh, do I as a user have to call `subsystem::subsystem_specific_log_reader_config::from_kafka_config()` everytime? :angry:) but I think it's incremental progress towards a better set of abstractions.

> We're still leaking abstractions all over, 

Agreed, let me take another look at some of the contained variables and try to de-leak them further from `kafka::log_reader_config`.

> we're still mutating and re-using model::offsets to refer to both kafka offsets and raft offsets depending on when you inspect the config.

Unfortunately this is true down to our fetch_config implementation in the Kafka layer.

https://github.com/redpanda-data/redpanda/blob/af5c7949e97aafd4ab062afaca939db1c734c989/src/v/kafka/server/handlers/fetch.h#L170-L183

nvartolomei commented:
I'd try creating the new log reader config where I want it, i.e. for level
zero and have a method that the caller can use to build one from the other.
This could be a better way at incremental improvement. Then cloud storage,
local storage could all do the same and at the end you remove the old code.

Local storage/local readers ideally should not have dependency the other
way around anyway. I.e. storage shouldn't have an arrow to //src/v/kafka.

Just an idea.

On Tue, Aug 12, 2025 at 1:43 PM Willem Kaufmann ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/v/cloud_topics/level_zero/reader/reader.h
> <https://github.com/redpanda-data/redpanda/pull/27213#discussion_r2269721294>
> :
>
> > @@ -20,6 +20,20 @@ class partition;
>  namespace experimental::cloud_topics {
>  class data_plane_api;
>
> +struct level_zero_log_reader_config {
> +    static level_zero_log_reader_config
>
> explicit constructor would also be a good option here, no qualms about
> changing construction away from this annoying-to-type static factory
> function.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/27213#discussion_r2269721294>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEETWPVJ46GZGR5C3GQ6FT3NHOQFAVCNFSM6AAAAACDU67WLKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZTCMJQGQZTENRSGI>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

WillemKauf commented:
Force push to:
* Remove static `from_kafka_reader_config()` functions, prefer `explicit` constructors
* Move non-Kafka state from `kafka::log_reader_config` to lower level reader configs
* Address other code review comments

dotnwat commented on src/v/storage/types.h:
```diff
@@ -211,11 +211,6 @@ struct append_result {
     friend std::ostream& operator<<(std::ostream& o, const append_result&);
 };
 
-using opt_abort_source_t
-  = std::optional<std::reference_wrapper<ss::abort_source>>;
-
-using opt_client_address_t = std::optional<model::client_address_t>;
```
i kind of think we should just nuke these typedefs. what's their purpose:
`model::opt_abort_source_t`
`std::optional<ss::abort_source>`

WillemKauf replied:
basically just code deduplication tree-wide. but I agree that they are weird typedefs to have and probably worth removing

dotnwat commented on src/v/raft/state_machine.cc:
```diff
@@ -158,8 +158,8 @@ ss::future<> state_machine::apply() {
                * we have to limit reading to the committed offset.
                */
-              storage::log_reader_config config(
-                _next, _raft->committed_offset());
+              kafka::log_reader_config kcfg(_next, _raft->committed_offset());
+              auto config = storage::local_log_reader_config(kcfg);
               return _raft->make_reader(config);
```
why is raft dealing in terms of kafka?

WillemKauf replied:
Hmm. It's a good point you raise.

Do you think there should be a (roughly) equivalent `model::log_reader_config` that is created from a `kafka::log_reader_config` and then instead encapsulate that by `storage::local_log_reader_config` et al?

andrwng replied:
Who else should use Raft offsets besides the local storage reader?

WillemKauf replied:
> Who else should use Raft offsets besides the local storage reader?

I suppose it is just the local storage reader who is interpreting these `model::offset`s as Raft offsets.

Besides adding another intermediary `model::local_log_reader_config`, do you see any good solution to this encapsulation problem?

andrwng replied:
I haven't looked at the call sites specifically so maybe this doesn't work, but does something like this work, or do the existing readers rely on Raft offsets too much?:

- storage::local_log_reader_config keeps model::offsets
- everything else uses kafka::log_reader_config, which have kafka::offsets
- one of the constructors for storage::local_log_reader_config (or the callers themselves) takes a kafka::log_reader_config and an offset_translator to do the conversion

WillemKauf replied:
> existing readers rely on Raft offsets too much

There are some low level fetch constructs in `kafka` that use `model::offset` instead of `kafka::offset`, annoyingly.

https://github.com/redpanda-data/redpanda/blob/af5c7949e97aafd4ab062afaca939db1c734c989/src/v/kafka/server/handlers/fetch.h#L170-L183

andrwng replied:
I don't quite see how that's problematic, given it's pretty far from the reader though? In these cases we can either:
- change them to use kafka::offset, because they _should_ be kafka offsets, or
- offset cast these to Kafka offsets when creating the kafka::log_reader_config

andrwng replied:
If the latter is sufficient, we can always work backwards from there in a separate PR to implement the former

WillemKauf replied:
Changed the PR to match something closer to this idea, in that:

* `kafka::log_reader_config` now operates on `kafka::offset`s instead of `model::offset`s.
* `storage::local_log_reader_config` can be created from a `kafka::log_reader_config` and an `offset_translator_state` object.
* `cloud_storage::cloud_log_reader_config` and `level_zero_log_reader_config` both contain a `kafka::log_reader_config`

I should still find a better way to split the last commit up, but its a lot of tree-wide changes that largely depend on each other.

andrwng commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -12,15 +12,17 @@
 struct log_reader_config {
-    model::offset start_offset;
-    model::offset max_offset;
+    kafka::offset start_offset;
+    kafka::offset max_offset;
```
Any reason for the kafka::offsets to not be in the commit that introduced the kafka::log_reader_config?

WillemKauf replied:
Nope. Squashed into the original commit

andrwng commented on src/v/kafka/data/replicated_partition.cc:
```diff
@@ -165,31 +166,34 @@ kafka::leader_epoch replicated_partition::leader_epoch() const {
-        cfg.type_filter = {model::record_batch_type::raft_data};
+      may_read_from_cloud(cfg.start_offset)
+      && cfg.start_offset
+           >= model::offset_cast(_partition->start_cloud_offset())) {
```
For posterity, this is indeed a Kafka offset... 🤦 

Once this change is in, the remaining offset casts will be very satisfying to remove

WillemKauf replied:
Yeah, I think we can work our way back down and up the tree now and hopefully get rid of most if not all offset casting within this path.

andrwng commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -476,12 +476,13 @@ class partition_record_batch_reader_impl final
-                    if (current_ko > _seg_reader->config().max_offset) {
+                      _seg_reader->config().kcfg.max_offset);
+                    if (
+                      model::offset_cast(current_ko)
```
nit: wondering if it makes sense to cast current_ko when it's declared?

WillemKauf replied:
Fixed

andrwng commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -1261,19 +1260,21 @@ remote_partition::timequery(storage::timequery_config cfg) {
     auto start_offset = std::max(
-      cfg.min_offset,
-      kafka::offset_cast(stm_manifest.full_log_start_kafka_offset().value()));
+      model::offset_cast(cfg.min_offset),
```
Just making sure, is there any change to behavior of timequery configs? Is entirely dependent on the context in which they're used?

WillemKauf replied:
No, the behavior will be the exact same, this is just residual type ugliness from our bad use of interchangeable `model::offset`/`kafka:offset` within the archival layer (and reader configs).

andrwng reviewed (commented):
Pretty much LGTM, just a couple nits. I looked through the remaining offset_cast calls and they seem like a reasonable starting point. Awesome work!

andrwng commented on src/v/cloud_topics/level_zero/reader/reader.h:
```diff
@@ -20,6 +20,22 @@ class partition;
 namespace experimental::cloud_topics {
 class data_plane_api;
 
+struct level_zero_log_reader_config {
```
nit: maybe pull this definition out into its own commit too

WillemKauf replied:
Pulled out definition into its own commit

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--57cb0fee-3261-4f0b-a1d2-3b39d6dcec2b-->
<!--TEST RESULTS-->
<!--eb7ada50-e6ab-46c8-b08e-0951b3695f5d-->
<!--TEST RESULTS-->
<!--7164da91-f89a-4587-8832-f3427d594d72-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70683">build#70683</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70683#0198a090-3625-4e00-bd39-0f75c6de4049 | FLAKY         | 18/21    | upstream reliability is '94.17218543046357'. current run reliability is '85.71428571428571'. drift is 8.4579 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70695">build#70695</a></summary>


| test_class      | test_method                 | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason   |
|:----------------|:----------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------|
| MasterTestSuite | test_replica_pair_frequency |                  | unit        | https://buildkite.com/redpanda/redpanda/builds/70695#0198a11d-1d71-4405-9d41-b283f7ab2430 | FAIL          | 0/1      |          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70816">build#70816</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70816#0198a9b9-5357-4aeb-982f-8fff0b12c520 | FLAKY         | 18/21    | upstream reliability is '93.63867684478372'. current run reliability is '85.71428571428571'. drift is 7.92439 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

WillemKauf committed:
`cloud_topics`: add `cloud_topic_log_reader_config`

WillemKauf commented:
To Nicolae's point about how lower level log readers should ideally avoid a dependency on `//src/v/kafka/data:log_reader_config`: would it be better to duplicate state into `cloud_log_reader_config` and `level_zero_log_reader_config` and adding similar functions to the newly added factory function `kafka::kafka_to_local_log_reader_config()` for these types, or is embedding `kafka::log_reader_config` into these objects and having the dependency an acceptable approach?

rockwotj commented:
> To Nicolae's point about how lower level log readers should ideally avoid a dependency on `//src/v/kafka/data:log_reader_config`: would it be better to duplicate state into `cloud_log_reader_config` and `level_zero_log_reader_config` and adding similar functions to the newly added factory function `kafka::kafka_to_local_log_reader_config()` for these types, or is embedding `kafka::log_reader_config` into these objects and having the dependency an acceptable approach?

I have yet to fully review (I will tomorrow!), but I'm infavor of just duplicating the state into the different subsystems: the translations are trivial, we have a simpler dependency graph, and it allows them to diverge in places where it makes sense more naturally. Maybe I need to do a full review first however.

andrwng commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,85 @@
+// since log_reader_config doesn't have a default constructor.
+inline const log_reader_config empty_reader_config{{}, {}};
+
+storage::local_log_reader_config kafka_to_local_log_reader_config(
+  log_reader_config, ss::lw_shared_ptr<const storage::offset_translator_state>);
```
This seems like the only reason kafka/data:log_reader_config depends on storage -- consider flipping the dependency considering how lightweight the log_reader_config is otherwise (basically only depends on model). Maybe move this constructor into storage somewhere. Otherwise it seems like a needlessly heavy dependency.

andrwng replied:
Alternatively, inline the calls. It's not a lot of code.

WillemKauf replied:
Flipped this dependency.

andrwng reviewed (commented):
>To Nicolae's point about how lower level log readers should ideally avoid a dependency on //src/v/kafka/data:log_reader_config

IMO if we make kafka/data:log_reader_config really lightweight (which, it's almost there), it doesn't feel like there's anything wrong with depending on it.

WillemKauf commented:
Force push to:
* Rebase after remove `include_prefix` [PR merge](https://github.com/redpanda-data/redpanda/pull/27226).

rockwotj reviewed (commented):
Alright, this is a great change. I need to double check a couple of the offset casts again in the morning, but trusting you and others have done due diligence here 😄 

My only comment is that I think the `kafka::log_reader_config` struct should either be:

* moved into the `model` namespace. It seems quite basic and universal, it seems better there (see later point).
* We should just duplicate the needed fields into the other reader config structs and write simple translation between them.

I don't have a strong feeling between either option, but would personally do the second. My reasoning being is that I'd like to keep our dependency graph cycle free at the *package* level. Bazel only enforces a DAG at the target level. Having a package level DAG I think makes the code base easier to understand, because each subsystem naturally will build on top of another. Otherwise one has to hop around the tree to understand the semantics of the subsystem.

See a related slack thread with Travis in `#p-bazel` on this subject: https://redpandadata.slack.com/archives/C0794GQ5NL9/p1754415704754519 and [a PR to do this with other packages](https://github.com/redpanda-data/redpanda/pull/27244)

WillemKauf commented:
>I need to double check a couple of the offset casts again in the morning, but trusting you and others have done due diligence here 

These casts should not affect correctness- if they were incorrect (e.g. if we were performing a comparison between a `kafka::offset` and a `model::offset` or passing a `kafka::offset` to a function expecting a `model::offset` parameter), the code would simply fail to compile.

> We should just duplicate the needed fields into the other reader config structs and write simple translation between them.

I agree, and I suppose this means we should port the existing `storage::kafka_to_local_log_reader_config()` function to (probably) somewhere within `kafka/replicated_partition`, and also add similar functions for `cloud_log_reader_config` and `level_zero_log_reader_config`. Does that line up with your vision for the dependency graph?

rockwotj commented:
> Does that line up with your vision for the dependency graph?

Yes

WillemKauf committed:
`kafka`: add conversion functions for `log_reader_config`s

WillemKauf committed:
`tree-wide`: remove dependencies on `storage::log_reader_config` throughout

Up till now, the `storage::log_reader_config` has been used as a way to specify
the desire to read from a Kafka log in `redpanda`. Unfortunately, this is a
leaky abstraction for a number of reasons, and also contains some local
`storage` specific flags that don't have any meaning outside of a `log_reader`
created over local `segment`s.

Despite this, `storage::log_reader_config` has been re-used in `cloud_storage`
code, as well as newly introduced `cloud_topics` code. This commit uses the newly
added definitions `storage::local_log_reader_config`, `cloud_storage::cloud_log_reader_config`,
and `cloud_topics::cloud_topic_log_reader_config` instead.

This would be a mess of a commit to try and split up, so its all in one big
tree-wide change.

WillemKauf commented on src/v/storage/types.h:
```diff
@@ -372,52 +410,15 @@ struct log_reader_config {
     // separately.
-    translate_offsets translate_offsets{false};
-
-    std::optional<ss::semaphore::clock::time_point> read_lock_deadline{};
```
this was unused, FYI.

oleiman replied:
ah, the deadline was used in an `async_data_uploader` PR that I'm reviving after several weeks on the back burner. fair enough removing dead code, anyway.

dotnwat replied:
hah, sorry. a new form of merge skew

rockwotj commented on src/v/kafka/data/log_reader_config.h:
```diff
@@ -0,0 +1,76 @@
+          as,
+          std::move(client_addr)) {}
+
+    friend std::ostream& operator<<(std::ostream& o, const log_reader_config&);
```
nit: consider using the new stuff in `base/format_to.h` instead of `operator<<`

WillemKauf replied:
Done

rockwotj commented on src/v/cloud_storage/types.h:
```diff
@@ -430,6 +430,80 @@ struct download_request {
+    // that the reader will return no batches.
+    bool strict_max_bytes{false};
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const cloud_log_reader_config&);
```
ditto comment on `format_to`

WillemKauf replied:
Done

rockwotj commented on src/v/cloud_topics/log_reader_config.h:
```diff
@@ -0,0 +1,95 @@
+
+    bool skip_cache{false};
+
+    friend std::ostream&
+    operator<<(std::ostream& o, const cloud_topic_log_reader_config&);
```
:smile:

WillemKauf replied:
:smiley_cat:

rockwotj commented on src/v/kafka/data/partition_proxy.h:
```diff
@@ -202,4 +205,14 @@ make_partition_proxy(const model::ktp&, cluster::partition_manager&);
 std::optional<partition_proxy>
 make_partition_proxy(const model::ntp&, cluster::partition_manager&);
 
+storage::local_log_reader_config kafka_to_local_log_reader_config(
+  log_reader_config cfg,
+  ss::lw_shared_ptr<const storage::offset_translator_state> ot_state);
+
+cloud_storage::cloud_log_reader_config
+kafka_to_cloud_log_reader_config(log_reader_config cfg);
+
+experimental::cloud_topics::cloud_topic_log_reader_config
+kafka_to_cloud_topic_log_reader_config(log_reader_config cfg);
```
Is this really the right spot for this? I suppose it's fine right now, but I would expect these to live closer to where there are needed, as I don't expect a user of `partition_proxy` to need these methods (an implementer yes, but not for example the kafka server).

I could justify it being in a header if we wanted to write tests for the conversions, but since we're not I'd prefer to hide these functions a bit (especially to not leak implementations of `partition_proxy` into this file).

rockwotj replied:
For example it looks like `kafka_to_cloud_log_reader_config` is only used in `replicated_partition.cc` so that can become private to that translation unit.

WillemKauf replied:
Moved these functions to anonymous name space  within `replicated_partition.cc` and `cloud_topic_partition.cc`

rockwotj commented on src/v/kafka/data/partition_proxy.h:
```diff
@@ -202,4 +205,14 @@ make_partition_proxy(const model::ktp&, cluster::partition_manager&);
+  ss::lw_shared_ptr<const storage::offset_translator_state> ot_state);
+
+cloud_storage::cloud_log_reader_config
+kafka_to_cloud_log_reader_config(log_reader_config cfg);
```
nit: consider calling `tiered_storage` to make it less similar to the `cloud_topic` version

WillemKauf replied:
I'd like to stick with the `cloud_log_reader_config` name, given that we have functions e.g `make_cloud_reader()`.

I think these definitions & uses are tightly scoped enough (and namespaced) that they shouldn't be confused for one another anywhere in the code.

rockwotj reviewed (commented):
LGTM, just a few nits on using `format_to` and hiding the translation functions better.

rockwotj commented on src/v/storage/types.cc:
```diff
@@ -45,7 +47,29 @@ std::optional<kafka::offset> stm_manager::lowest_pinned_data_offset() const {
+      client_address);
+}
+
+std::ostream& operator<<(std::ostream& o, const local_log_reader_config& cfg) {
```
nit: can be removed, as `format_to` provides this function, but I see the declaration is removed from the header, so is dead code now?

WillemKauf replied:
Oops, missed this removal. Pushed.

andrwng commented on src/v/kafka/data/replicated_partition.cc:
```diff
@@ -165,31 +204,34 @@ kafka::leader_epoch replicated_partition::leader_epoch() const {
+      && cfg.start_offset
+           >= model::offset_cast(_partition->start_cloud_offset())) {
+        auto config = kafka_to_cloud_log_reader_config(cfg);
+        config.type_filter = {model::record_batch_type::raft_data};
```
I might just be missing it, but noting that the cloud storage reader config type filter seems like it isn't used meaningfully. Seems like it's only ever raft_data and used when converting to a storage log config, in which case we could hard code that at the translation sites

WillemKauf replied:
Agreed, the type filter doesn't seem to provide any value in the `cloud_log_reader_config`, we could look at removing it in a follow up.

dotnwat commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -471,13 +469,13 @@ class partition_record_batch_reader_impl final
                 } else {
-                    auto current_ko = _ot_state->from_log_offset(
-                      _seg_reader->current_rp_offset());
+                    auto current_ko = model::offset_cast(
```
so before current_ko was model::offset, and now it is kafka::offset? that makes sense, but what is confusing is why no more changes below are needed. it seems like that might suggest that named_type isn't providing strong enough type guarnatees?

WillemKauf replied:
> so before current_ko was model::offset, and now it is kafka::offset?

Correct.

> that makes sense, but what is confusing is why no more changes below are needed. it seems like that might suggest that named_type isn't providing strong enough type guarnatees?

No, the type guarantees are correct, the difference is the offsets being compared from within the `cloud_log_reader_config` are now _also_ `kafka::offset`s instead of `model::offset`s (subtle)

dotnwat commented on src/v/cloud_storage/remote_partition.cc:
```diff
@@ -1261,13 +1257,14 @@ remote_partition::timequery(storage::timequery_config cfg) {
+      stm_manifest.full_log_start_kafka_offset().value());
+    auto max_offset = model::offset_cast(cfg.max_offset);
 
     // Synthesize a log_reader_config from our timequery_config
-    storage::log_reader_config config(
+    cloud_log_reader_config config(
       start_offset,
-      cfg.max_offset,
+      max_offset,
```
one this this PR is highlighting is that it seems the model/kafka type difference isn't being used to its full potential if we can just cast the types around whenever??

WillemKauf replied:
> one this this PR is highlighting is that it seems the model/kafka type difference isn't being used to its full potential if we can just cast the types around whenever??

Absolutely, I keep highlighting the fact that down low in the `fetch` path, the `fetch_config` is using `model::offset`s to represent `kafka::offset`s for whatever reason. 


https://github.com/redpanda-data/redpanda/blob/af5c7949e97aafd4ab062afaca939db1c734c989/src/v/kafka/server/handlers/fetch.h#L170-L183

We should really tighten up our use of the cast & better respect offset semantics tree-wide, but it may have to start from this lower level.

dotnwat replied:
> We should really tighten up our use of the cast & better respect offset semantics tree-wide, but it may have to start from this lower level.

yeh. good follow up work. i could see it working either top-down or bottom-up or middle-out

dotnwat commented on src/v/kafka/data/rpc/test/deps.h:
```diff
@@ -110,23 +110,22 @@ class in_memory_proxy : public kafka::partition_proxy::impl {
-        if (
-          config.first_timestamp.has_value()
-          || config.type_filter.has_value()) {
+        if (config.first_timestamp.has_value()) {
```
why is hte type filter gone?

WillemKauf replied:
Because the `kafka::log_reader_config` doesn't have a type filter anymore.

dotnwat commented on src/v/storage/lock_manager.cc:
```diff
@@ -96,9 +96,7 @@ lock_manager::range_lock(const log_reader_config& cfg) {
-    return range(
-      std::move(tmp),
-      cfg.read_lock_deadline.value_or(ss::semaphore::clock::time_point::max()));
+    return range(std::move(tmp));
```
why is deadline gone?

dotnwat commented:
@WillemKauf all good merging the PR looks good. can you quickly check out my comments and andrew's comment from a few hours ago?
